pkg/kubelet/kubelet_pods.go

```diff
@@ -675,6 +676,38 @@ func (kl *Kubelet) podIsTerminated(pod *v1.Pod) bool {
 	return false
 }
 
+func (kl *Kubelet) safeToDeletePod(pod *v1.Pod) bool {
+	if kubepod.IsMirrorPod(pod) {
+		// We don't handle graceful deletion of mirror pods.
+		return false
+	}
+	if pod.DeletionTimestamp == nil {
+		// We shouldnt delete pods whose DeletionTimestamp is not set
+		return false
+	}
+	if !notRunning(pod.Status.ContainerStatuses) {
+		// We shouldnt delete pods that still have running containers
+		glog.V(3).Infof("Pod %q is terminated, but some containers are still running", format.Pod(pod))
+		return false
+	}
+	if kl.podVolumesExist(pod.UID) && !kl.kubeletConfiguration.KeepTerminatedPodVolumes {
+		// We shouldnt delete pods whose volumes have not been cleaned up if we are not keeping terminated pod volumes
```

Comment:
There is logging related to volume unmounting in [pkg/kubelet/volumemanager/reconciler/reconciler.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go).  My only concern is that most debugging people wont make the connection that their pod is not being deleted because unmounting a volume failed.  But I'm not sure how to notify users without logging every time they try to delete a pod (which is useless).